src/trs/rewrite.rs

```diff
@@ -0,0 +1,143 @@
+//! Polymorphically typed (Hindley-Milner) First-Order Term Rewriting Systems (no abstraction)
+//!
+//! Much thanks to:
+//! - https://github.com/rob-smallshire/hindley-milner-python
+//! - https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system
+//! - (TAPL; Pierce, 2002, ch. 22)
+
+use polytype::Context as TypeContext;
+use rand::seq::sample_iter;
+use rand::Rng;
+use std::f64::NEG_INFINITY;
+use std::fmt;
+use term_rewriting::trace::Trace;
+use term_rewriting::{Rule, RuleContext, TRS as UntypedTRS};
+
+use super::{Lexicon, ModelParams, SampleError, TypeError};
+
+/// Manages the semantics of a term rewriting system.
+#[derive(Debug, PartialEq, Clone)]
+pub struct TRS {
+    // TODO: may also want to track background knowledge here.
+    pub(crate) lex: Lexicon,
+    // INVARIANT: UntypedTRS.rules ends with lex.background
+    pub(crate) utrs: UntypedTRS,
+    pub(crate) ctx: TypeContext,
+}
+impl TRS {
+    /// Create a new `TRS` under the given `Lexicon`. Any background knowledge will be appended to
+    /// the given ruleset.
+    pub fn new(lex: &Lexicon, mut rules: Vec<Rule>) -> Result<TRS, TypeError> {
+        let lex = lex.clone();
+        let mut ctx = TypeContext::default();
+        let utrs = {
+            let lex = lex.0.read().expect("poisoned lexicon");
+            rules.append(&mut lex.background.clone());
+            let utrs = UntypedTRS::new(rules);
+            lex.infer_utrs(&utrs, &mut ctx)?;
+            utrs
+        };
+        Ok(TRS { lex, utrs, ctx })
+    }
+
+    /// The size of the TRS (the sum over the size of the rules in the underlying [`TRS`])
+    ///
+    /// [`TRS`]: ../../term_rewriting/struct.TRS.html
```

Comment:
Use docs.rs link, and say `term_rewriting::TRS` to make it clear that it's in another crate.

Maybe change the docs to simply say "The size of the underlying `term_rewriting::TRS`", and have that link to the method docs at `https://docs.rs/term_rewriting/~0.3/term_rewriting/struct.TRS.html#method.size`